art/attacks/evasion/simba.py

```diff
@@ -0,0 +1,319 @@
+                             + str(classifier.__class__.__bases__) + '. '
+                             ' The classifier needs to be a Neural Network and provide gradients.'))
+
+        params = {'attack': attack, 'max_iter': max_iter, 'epsilon': epsilon, 'order': order, 'freq_dim': freq_dim, 'stride': stride, 'targeted': targeted ,'batch_size': batch_size}
+        self.set_params(**params)
+
+    def generate(self, x, y=None, **kwargs):
+        """
+        Generate adversarial samples and return them in an array.
+
+        :param x: An array with the original inputs to be attacked.
+        :type x: `np.ndarray`
+        :param y: An array with the original labels to be predicted.
+        :type y: `np.ndarray`
+        :return: An array holding the adversarial examples.
+        :rtype: `np.ndarray`
+        """
+        x = x.astype(ART_NUMPY_DTYPE)
+        preds = self.classifier.predict(x, batch_size=self.batch_size)
+
+        if y is None:
+            if self.targeted:
+                raise ValueError('Target labels `y` need to be provided for a targeted attack.')
+            else:
+                # Use model predictions as correct outputs
+                logger.info('Using the model prediction as the correct label for SimBA.')
+                y_i = np.argmax(preds, axis=1)
+        else:
+            y_i = np.argmax(y, axis=1)
+        
+        desired_label = y_i[0]
+        current_label = np.argmax(preds, axis=1)[0]
+        last_prob = preds.reshape(-1)[desired_label]
+
+        n_dims = np.prod(x.shape)
+
+        if self.attack == 'px':
+            if self.order == 'diag':
+                indices = self.diagonal_order(x.shape[2], 3)[:self.max_iter]
+            elif self.order == 'random':
+                indices = np.random.permutation(n_dims)[:self.max_iter]
+            indices_size = len(indices)
+            while indices_size < self.max_iter:
+                if self.order == 'diag':
+                    tmp_indices = self.diagonal_order(x.shape[2], 3)
+                elif self.order == 'random':
+                    tmp_indices = np.random.permutation(n_dims)
+                indices = np.hstack((indices, tmp_indices))[:self.max_iter]
+                indices_size = len(indices)
+        elif self.attack == 'dct':
+            indices = self._block_order(x.shape[2], 3, initial_size=self.freq_dim, stride=self.stride)[:self.max_iter]
+            indices_size = len(indices)
+            while indices_size < self.max_iter:
+                tmp_indices = self._block_order(x.shape[2], 3, initial_size=self.freq_dim, stride=self.stride)
+                indices = np.hstack((indices, tmp_indices))[:self.max_iter]
+                indices_size = len(indices)
+            trans = lambda z: self._block_idct(z, block_size=x.shape[2])
+
+        clip_min = -np.inf
+        clip_max = np.inf 
+        if hasattr(self.classifier, 'clip_values') and self.classifier.clip_values is not None:
```

Comment:
`clip_values` is now a property.
```suggestion
        if self.classifier.clip_values is not None:
```